### DISCUSSION TITLE
What is underscore in terms of real or floating point fragments

### DISCUSSION THREAD

**Discussion by mwpowellhtx** [antlr/antlr4#4649] [Q&A]
Referencing in particular the [_CSharp_ `REAL_LITERAL`](https://github.com/antlr/grammars-v4/blob/master/csharp/CSharpLexer.g4#L147C1-L149C62) lexicon, perhaps others have modeled similarly:

```antlr
REAL_LITERAL:
    ([0-9] ('_'* [0-9])*)? '.' [0-9] ('_'* [0-9])* ExponentPart? [FfDdMm]?
    | [0-9] ('_'* [0-9])* ([FfDdMm] | ExponentPart [FfDdMm]?)
```

Mainly am interested in a subset portion of the definition:

```antlr
DECIMAL_LITERAL:
    ([0-9] ('_'* [0-9])*)? '.' [0-9] ('_'* [0-9])*
```

However, what is the significance of the `'_'` in the definition, I am unfamiliar with that syntax.

As if to say, could be something like this? `0___0`, really? After a bit more digging, yes, truly, I guess so; if I am [_keeping with that definition_](https://github.com/dotnet/roslyn/issues/216).

No harm if my particular use cases do not need it? Otherwise, the lexicon seems pretty solid. Better than what I was tinkering with, fewer alternatives that can trip up and foul things up:

```antlr
DECIMAL_LITERAL: MINUS? ( ( DIGIT+ DOT? ) | ( DIGIT+? DOT DIGIT+ ) );
```

Thoughts, suggestions? Other than of course to pare back the _CSharp_ description a bit and just run with it.

**Comment by mwpowellhtx** [comment]
To clarify a bit, basically, I want to support the following kinds of use cases:

```
1.2
-1.2
.2
-.2
1.
-1.
```

Maybe that's asking too much. i.e. just define an `INTEGER_LITERAL` and be done with it. The decimal part being, when there is a DOT involved, then it is decimal.

**Comment by mwpowellhtx** [comment]
Having a devil of a time making this work correctly. Trying to get this "simple" DSL working, but it cannot seem to parse beyond the first `coordinate`, if it gets that far.

```antlr
fragment DIGIT: [0-9];
fragment DOT: '.';
fragment MINUS: '-';
DECIMAL_LITERAL: DIGIT+ ( DOT DIGIT+ )?;
NUMERIC_LITERAL: MINUS? DECIMAL_LITERAL;

SP: ' ';
WS: ( SP | HT | VT )+ -> skip;

location: coordinate ( SP+ coordinate )*;
coordinate: NUMERIC_LITERAL;
```

I've tried different iterations of this, none of them seem to be quite working. For now focused there. Goal is to support the decimal forms above. But like I said, I am willing to consider an `INTEGER | DECIMAL` style approach if that works better.

Eventually in its slightly more complex axis based form:

```antlr
// Importing the baseline grammar
fragment X: [xX];
fragment Y: [yY];
fragment Z: [zZ];
AXIS: X | Y | Z;
OP_EQ: '=';

coordinate: AXIS SP* OP_EQ SP* DECIMAL_LITERAL;
```

**Comment by mwpowellhtx** [comment]
By comparison, or perhaps contrast at this point, the same problem approached from a regular expression perspective has been proven far more reliable.

```csharp
decimal val;
var re = new Regex(@$"\s*(?<{nameof(val)}>\-?(\d+(\.\d*)?|\d*\.\d+))");
// And subsequent logic parsing through the matches etc.
```

**Reply by kaby76** [reply]
```
SP: ' ';
WS: ( SP | HT | VT )+ -> skip;
```
You cannot define token SP and then token WS following it. Antlr lexers follow two rules:
* Match the token to the longest string first, regardless of the order of the lexer rules.
* If two or more lexer rules match a string of the same length, then match the first lexer rule.

Given that, a space will match SP, never WS. While both SP and WS match a single space character, the first rule--SP--will always be selected.

Also: Antlr parsers never "guide" the lexer. You can think of an Antlr lexer tokenizing the entire input before the parser does anything.

------

In C#, '_' means nothing as far as the number, but it is included for ["readability"](https://stackoverflow.com/questions/43476056/what-do-the-underscores-mean-in-a-numeric-literal-in-c). So, this is valid C#:
```
var r = 1.123_456_789;
var r2 = 1.123456789; // the same as r.
Console.WriteLine(r);
```
----

`NUMERIC_LITERAL: MINUS? DECIMAL_LITERAL;` For a general programming language, you should define the syntax of a number separately from the unary operator '-' or '+'. In other words, do not include the '-'-sign with the lexer rule for the syntax of a number.

----

`By comparison, or perhaps contrast at this point, the same problem approached from a regular expression perspective has been proven far more reliable.` The right-hand side of an Antlr lexer rule *is* a regular expression. But, you need to remember the two rules of how Antlr lexers "work".

**Reply by mwpowellhtx** [reply]
Ah... I was wondering whether maybe whitespace was being gobbled up when I didn't want it to be. I'll try those suggestoins, thanks...

**Reply by mwpowellhtx** [reply]
Following up, I tried adjusting the lexer and rules, still no luck. It is not matching even the first coordinate attribute.

```antlr
// Lexer rules
fragment DIGIT: [0-9];
fragment DOT: '.';
fragment MINUS: '-';
fragment DECIMAL_LITERAL: DIGIT+ ( DOT DIGIT+ )?;

WS: ( ' ' | '\t' | '\f' )+ -> skip;

// Grammar rules
location: coordinate+;

coordinate: numericLiteral;

numericLiteral: MINUS DECIMAL_LITERAL | DECIMAL_LITERAL;
```

And the _axis_ based grammar, _"extending"_ only coordinate.

```antlr
fragment X: [xX];
fragment Y: [yY];
fragment Z: [zZ];

AXIS: X | Y | Z;

OP_EQ: '=';

coordinate: AXIS OP_EQ numericLiteral;
```

I'm a bit confused, when to or not to `fragment`, could that be influencing the results, i.e. when I just want a simple tokenizer, such as this.

Under test, it's not even reaching the first coordinate, with breakpoints set in my:

```csharp
public override void ExitCoordinate([NotNull] ScumLocationParser.CoordinateContext context) { ... }
```

So something's wrong, but I'm not sure exactly what.

**Reply by kaby76** [reply]
* `fragment` is used on a lexer rule if the rule is not supposed to create a token. It's used when you want to reuse some expression across multiple lexer rules, or to just make a rule look "nice".
* Lexer symbols that are defined with `fragment` cannot be used in a parser rule because there is no token produced by the lexer. So, `fragment DECIMAL_LITERAL: DIGIT+ ( DOT DIGIT+ )?; ... numericLiteral: MINUS DECIMAL_LITERAL | DECIMAL_LITERAL;` cannot work because `DECIMAL_LITERAL` is not a normal symbol, having no token value. It cannot be produced by the lexer so `numericLiteral` can never match. Also, you should write the rule `numericLiteral : MINUS? DECIMAL_LITERAL;`, not `numericLiteral: MINUS DECIMAL_LITERAL | DECIMAL_LITERAL;`. Use Antlr `?`-operators.
* Before running the parser, you should just write a driver program that outputs the tokens, using `NextToken()`, or `nextToken()`.

**Reply by mwpowellhtx** [reply]
Thinking `fragment` is having more of an adverse effect than we're crediting.

I tried the `?` operator, still does not work. So it's not that, not exactly, anyway.

Put breakpoints in for Enter/Exit Every Rule. Enter, then Exit hit; EOF. It is not even recognizing the first decimal numeric literal.

Context showing `source` is `{(WhalleyBotEnhanced.ScumLocationLexer,     -.4    .4    -.4    .45)}`, `(start, stop)` is `(27, 26)`.

**Reply by kaby76** [reply]
* You don't have an EOF-terminated start rule. Antlr does not work (well) using a grammar that is clearly a start rule. Something like `start: location EOF;`.

**Comment by kaby76** [comment]
Not sure what you are doing wrong. Here's a zip with driver, grammar, etc., for C#. 
[Generated-CSharp.zip](https://github.com/user-attachments/files/16041890/Generated-CSharp.zip)


```
06/29-14:52:32 ~/tes/yyy
$ ls
G.g4
06/29-14:52:34 ~/tes/yyy
$ cat G.g4
grammar G;

// Lexer rules
fragment DIGIT: [0-9];
fragment DOT: '.';
MINUS: '-';
DECIMAL_LITERAL: DIGIT+ ( DOT DIGIT+ )?;
fragment X: [xX];
fragment Y: [yY];
fragment Z: [zZ];
AXIS: X | Y | Z;
OP_EQ: '=';
WS: ( ' ' | '\t' | '\f' )+ -> skip;

// Grammar rules
location: coordinate+;
coordinate: AXIS OP_EQ numericLiteral;
numericLiteral: MINUS DECIMAL_LITERAL | DECIMAL_LITERAL;
start: location EOF;
06/29-14:52:36 ~/tes/yyy
$ trgen -t CSharp
CSharp  G.g4 success 0.0366777
Rendering template file from CSharp/Other.csproj to ./Generated-CSharp/Other.csproj
Rendering template file from CSharp/st.build.ps1 to ./Generated-CSharp/st.build.ps1
Rendering template file from CSharp/st.build.sh to ./Generated-CSharp/st.build.sh
Rendering template file from CSharp/st.clean.ps1 to ./Generated-CSharp/st.clean.ps1
Rendering template file from CSharp/st.clean.sh to ./Generated-CSharp/st.clean.sh
Rendering template file from CSharp/st.Encodings.cs to ./Generated-CSharp/st.Encodings.cs
Rendering template file from CSharp/st.ErrorListener.cs to ./Generated-CSharp/st.ErrorListener.cs
Rendering template file from CSharp/st.makefile to ./Generated-CSharp/st.makefile
Rendering template file from CSharp/st.perf.sh to ./Generated-CSharp/st.perf.sh
Rendering template file from CSharp/st.ProfilingCommonTokenStream.cs to ./Generated-CSharp/st.ProfilingCommonTokenStream.cs
Rendering template file from CSharp/st.run.ps1 to ./Generated-CSharp/st.run.ps1
Rendering template file from CSharp/st.run.sh to ./Generated-CSharp/st.run.sh
Rendering template file from CSharp/st.test-cover.sh to ./Generated-CSharp/st.test-cover.sh
Rendering template file from CSharp/st.Test.cs to ./Generated-CSharp/st.Test.cs
Rendering template file from CSharp/st.test.ps1 to ./Generated-CSharp/st.test.ps1
Rendering template file from CSharp/st.test.sh to ./Generated-CSharp/st.test.sh
Rendering template file from CSharp/Test.csproj.st to ./Generated-CSharp/Test.csproj.st
Copying source file from C:/msys64/home/Kenne/tes/yyy/G.g4 to ./Generated-CSharp/G.g4
06/29-14:52:42 ~/tes/yyy
$ cd Generated-CSharp/
06/29-14:52:45 ~/tes/yyy/Generated-CSharp
$ make
bash build.sh
  Determining projects to restore...
  Restored C:\msys64\home\Kenne\tes\yyy\Generated-CSharp\Test.csproj (in 509 ms).
  Determining projects to restore...
  All projects are up-to-date for restore.
  Test -> C:\msys64\home\Kenne\tes\yyy\Generated-CSharp\bin\Debug\net8.0\Test.dll

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:03.24
06/29-14:52:51 ~/tes/yyy/Generated-CSharp
$ ./bin/Debug/net8.0/Test.exe -input "x=1.1 y=2.2 z=3.3" -tokens
[@0,0:0='x',<3>,1:0]
[@1,1:1='=',<4>,1:1]
[@2,2:4='1.1',<2>,1:2]
[@3,6:6='y',<3>,1:6]
[@4,7:7='=',<4>,1:7]
[@5,8:10='2.2',<2>,1:8]
[@6,12:12='z',<3>,1:12]
[@7,13:13='=',<4>,1:13]
[@8,14:16='3.3',<2>,1:14]
[@9,17:16='<EOF>',<-1>,1:17]

CSharp 0 string0 success 0.0058308
Total Time: 0.0575644
06/29-14:53:05 ~/tes/yyy/Generated-CSharp
$
```

**Reply by mwpowellhtx** [reply]
Definitely not accustomed to running command line anything, and the scaffold I have is already established. Doing the best I can running through the listeners and such; although perhaps maybe could add more diagnostics there, but that aside.

Tried a couple of adjustments, definitely `fragment` is having an effect, although I am not positive exactly what. Difference between, in this case, signage being rolled up part of the _coordinate numeric text_, and not.

```antlr
MINUS: '-';
DECIMAL_LITERAL: DIGIT* DOT DIGIT+ | DIGIT+ ( DOT DIGIT* )?;
// ...
numeric: MINUS? DECIMAL_LITERAL;
```

Also reverted to my original _decimal definition_, as I think this is more reflective of the actual potential use cases.

Closer, I think, the results indicate as much. Now all value only coordinates use cases passing. I can troubleshoot the axis based cases; probably something similar going on there, something whether it should be `fragment` or not.

**Reply by mwpowellhtx** [reply]
Okay, and verified _axis listeners_, now having victory in both areas. 🎆 🍾 

And here I was ready to explore the possitibility of a _SWIG C++ Boost Spirit X3_ and/or _Qi_ based approach, big fan of that technology. However, there's a lot more integration work to do around that possibility extending into a dotnet assembly. 🐑 😬

**Reply by kaby76** [reply]
C++/SWIG is tough to deal with sometimes. Antlr with the C# target and Antlr4BuildTasks is much easier. Make sure to use antlr.runtime.standard, not the old Antlr4cs.

**Reply by mwpowellhtx** [reply]
Yeah I know. In and of itself, I was _"ready"_ with a templated X3 parser. But getting this to work was a huge relief. Good to know re: `standard`, thanks, and thanks for the assist.
